### ISSUE TITLE
Styling

### ISSUE THREAD

**Issue by ogoffart** [sixtyfpsui/sixtyfps#45]
We want the ability to

 * Change the style globably for all widgets
 * Make it easy to write your own style which is applied globaly
 * Change the style of particular widgets

**Comment by bitdisaster** [comment]
I think it would be important to make the UI look native to the default design language on each platform with built-in light and dark theme support.

**Comment by syf20020816** [comment]
```
			  default
————————————————————————————————————
|  logic control layer (Rust|C++)  |
————————————————————————————————————
				⇕
————————————————————————————————————
|    UI layer (write components)   |
————————————————————————————————————

		   SurrealismUI
————————————————————————————————————
|  logic control layer (Rust|C++)  |
————————————————————————————————————
				⇕
————————————————————————————————————
|      UI Styles Wrapper layer     |   <-- What SurrealismUI do , see ①
————————————————————————————————————
|   UI layer (write components)    |
————————————————————————————————————


				System support (like iced)
————————————————————————————————————      ————————————————
|           logic control          | -->  | Theme::Light |
————————————————————————————————————      ————————————————
|             UI layer             |     		  ↓
———————————————————————————————————— 	    |————————————|
						 ↑			     ↓            ↓
				import	  ← Light_Theme Styles   Dark_Theme Styles
```
## ①
define a lot replaceable theme styles and binding styles use theme property , can be customized in slint file or logic control layer , means: all system components are wrapped (Customizing themes in third-party component libraries is very affordable as it acts on the UI layer. SLINT is like an integration of HTML and CSS, so I use this way)(By binding global singleton variables to styles, any component that uses variables can change styles simultaneously)

## Diff
Slint differs from other GUI frameworks in that the UI layer is completed through. slint, which I believe is good and brings many advantages (compatibility with different platforms, instant preview, maintainability, parallel development, etc.). But this also leads to SLIT being unable to easily customize the theme of the component. Theme customization and switching are dynamic to static processes, which require a lot of logical processing, and this is also same as (HTML+CSS+js | ts)
## Slint please be careful
Slint's work on topic definition will simultaneously affect the built-in components currently provided and other languages' API. Although this feature may seem simple, it may bring significant risks, which I believe need to be weighed and considered. Is this necessary? Because for third-party component libraries, although the workload of helping users define themes is large, it is not complex. As the author of SurrealismUI, it is evident that we have successfully implemented the definition of themes in static slint language and can be modified at the logical level. Users customize themes through static file overwriting.

**Comment by Anivie** [comment]
As of this writing, Slint developers are still unable to style their projects, which is very detrimental to the product's artwork.

**Comment by ogoffart** [comment]
If you want custom style for your application, the current way to do it is to write your own control with your own style.
For example, if you want a custom looking button, you would write a Button which you would import.
There are some example in the docs in our recipes:  https://slint.dev/releases/1.2.0/docs/slint/src/recipes/recipes#custom-widgets.
Some of our example do that, for example the printerdemo has a custom style.

**Comment by Anivie** [comment]
> If you want custom style for your application, the current way to do it is to write your own control with your own style. For example, if you want a custom looking button, you would write a Button which you would import. There are some example in the docs in our recipes: https://slint.dev/releases/1.2.0/docs/slint/src/recipes/recipes#custom-widgets. Some of our example do that, for example the printerdemo has a custom style.

Yes, I saw that section. But actually creating widgets from 2D graphics is still too complicated, I'd rather have an api that allow us to write a custom-style for built-in components and migrate over to them as easily as the built-in styles.

**Comment by heipiao233** [comment]
I found that this works:
1. Create `include/xxstyle`
2. Implement std widgets in `xxstyle`
3. Create link `include/xxstyle-{dark,light}` to `include/xxstyle`
4. Add `include` and `include/xxstyle` to include path
5. Set style to `xxstyle`
However, if you add these folders to include path, built-in styles doesn't work. So we should only add them when using the custom style. I don't know whether this is supported.

**Comment by Baerbeisser** [comment]
> I think it would be important to make the UI look native to the default design language on each platform with built-in light and dark theme support.

On Linux, this means using the currently active Gtk/Qt/Elementary/... theme. Which is usually a hodgepodge of default and application-specific tag=color definitions, not unlike CSS. What does Slint currently do here? Using Adwaita? I'm using [Materia](https://github.com/nana-4/materia-theme/raw/master/images/widget-factory.png?raw=true) tho.
